include/Dialects/LinalgTransform/TransformOpInterface.h

```diff
@@ -0,0 +1,226 @@
+//===-- TransformOpInterface.h - Interface for transform ops ----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_LINALG_TRANSFORM_TRANSFORM_OP_INTERFACE_H
+#define MLIR_DIALECT_LINALG_TRANSFORM_TRANSFORM_OP_INTERFACE_H
+
+#include "Dialects/LinalgTransform/TransformOpMapping.h"
+#include "Transforms/Functional.h"
+
+namespace mlir {
+namespace linalg {
+namespace transform {
+
+class TransformOpInterface;
+
+/// The state maintained across applications of various ops implementing the
+/// TransformOpInterface. In particular, this contains the mapping between
+/// values defined by these ops in the transform IR (the part of the IR
+/// containing the ops that implement TransformOpInterface and guide the
+/// transformation) and payload IR (the part of the IR that is being
+/// transformed) ops. Assumes that each value in the transform IR can be used at
+/// most once (since transformations are likely to change the payload IR ops the
+/// value corresponds to). Checks that transform IR values correspond to
+/// disjoint sets of payload IR ops throughout the transformation.
+class TransformState {
+public:
+  /// Creates a state for the transformation rooted at the given op.
```

Comment:
Doesn't seem to be the case in MLIR...

Reply:
depends where you look, that's why I use "trying"; e.g. randomly picking on stuff:

mlir/Rewrite/FrozenRewritePatternSet.h does
mlir/Dialect/Vector/Transforms/VectorTransforms.h partially does
mlir/Dialect/Linalg/Transforms/Transforms.h does
mlir/Dialect/Affine/LoopFusionUtils.h is 50-50 ...

My recollection is this was an actual debate with somewhat strong opinions involved.

Reply:
It never made it to the coding guidelines so I suppose there is no limitation one way or another. I usually avoid nitpicking on these in reviews to avoid resuscitating that bikeshed discussion. If you do feel strongly about these, I can change the comments.